xapian-core/weight/tfidfweight.cc

```diff
@@ -36,85 +36,68 @@ using namespace std;
 {
-    if (normals.length() != 3 ||
-	!strchr("nbslPL", normals[0]) ||
-	!strchr("ntpfsP", normals[1]) ||
-	!strchr("n", normals[2]))
-	throw Xapian::InvalidArgumentError("Normalization string is invalid");
-    if (param_slope <= 0)
-	throw Xapian::InvalidArgumentError("Parameter slope is invalid");
-    if (param_delta <= 0)
-	throw Xapian::InvalidArgumentError("Parameter delta is invalid");
-    if (normals[1] != 'n') {
-	need_stat(TERMFREQ);
-	need_stat(COLLECTION_SIZE);
-    }
-    need_stat(WDF);
-    need_stat(WDF_MAX);
-    need_stat(WQF);
-    if (normals[0] == 'P' || normals[1] == 'P') {
-	need_stat(AVERAGE_LENGTH);
-	need_stat(DOC_LENGTH);
-	need_stat(DOC_LENGTH_MIN);
-    }
-    if (normals[0] == 'L') {
-	need_stat(DOC_LENGTH);
-	need_stat(DOC_LENGTH_MIN);
-	need_stat(DOC_LENGTH_MAX);
-	need_stat(UNIQUE_TERMS);
-    }
-    switch (normals[0]) {
+    switch (c) {
 	case 'b':
-	    wdf_norm_ = wdf_norm::BOOLEAN;
-	    break;
+	    return wdf_norm::BOOLEAN;
 	case 's':
-	    wdf_norm_ = wdf_norm::SQUARE;
-	    break;
+	    return wdf_norm::SQUARE;
 	case 'l':
-	    wdf_norm_ = wdf_norm::LOG;
-	    break;
+	    return wdf_norm::LOG;
 	case 'P':
-	    wdf_norm_ = wdf_norm::PIVOTED;
-	    break;
+	    return wdf_norm::PIVOTED;
 	case 'L':
-	    wdf_norm_ = wdf_norm::LOG_AVERAGE;
-	    break;
+	    return wdf_norm::LOG_AVERAGE;
+	case 'n':
+	    return wdf_norm::NONE;
 	default:
-	    wdf_norm_ = wdf_norm::NONE;
+	    throw Xapian::InvalidArgumentError
+		  ("Normalization string is invalid");
     }
-    switch (normals[1]) {
+}
+
+TfIdfWeight::idf_norm
+TfIdfWeight::decode_idf_norm(char c)
+{
+    switch (c) {
 	case 'n':
-	    idf_norm_ = idf_norm::NONE;
-	    break;
+	    return idf_norm::NONE;
 	case 's':
-	    idf_norm_ = idf_norm::SQUARE;
-	    break;
+	    return idf_norm::SQUARE;
 	case 'f':
-	    idf_norm_ = idf_norm::FREQ;
-	    break;
+	    return idf_norm::FREQ;
 	case 'P':
-	    idf_norm_ = idf_norm::PIVOTED;
-	    break;
+	    return idf_norm::PIVOTED;
 	case 'p':
-	    idf_norm_ = idf_norm::PROB;
-	    break;
+	    return idf_norm::PROB;
+	case 't':
+	    return idf_norm::TFIDF;
 	default:
-	    idf_norm_ = idf_norm::TFIDF;
+	    throw Xapian::InvalidArgumentError
+		  ("Normalization string is invalid");
     }
-    wt_norm_ = wt_norm::NONE;
 }
 
-TfIdfWeight::TfIdfWeight(wdf_norm wdf_normalization,
-			 idf_norm idf_normalization,
-			 wt_norm wt_normalization)
-    : TfIdfWeight::TfIdfWeight(wdf_normalization, idf_normalization,
-			       wt_normalization, 0.2, 1.0) {}
+TfIdfWeight::wt_norm
+TfIdfWeight::decode_wt_norm(char c)
+{
+    switch (c) {
+	case 'n':
+	    return wt_norm::NONE;
+	default:
+	    throw Xapian::InvalidArgumentError
+		  ("Normalization string is invalid");
+    }
+}
+
+TfIdfWeight::TfIdfWeight(const std::string& normals,
+			 double slope, double delta)
+    : TfIdfWeight::TfIdfWeight(decode_wdf_norm(normals[0]),
+			       decode_idf_norm(normals[1]),
+			       decode_wt_norm(normals[2]),
```

Comment:
There's no longer a check on the length of `normals` before we access its characters - pass in an empty string and this will read off the end of it.  That's why in my sketch of the code I passed `normals` to these helpers - then they can check `normals.size() == 3`.  We probably need to repeat that check in each helper since the order in which parameters are evaluated isn't specified in C++, but these functions are only called once so the compiler is likely to decide to inline them all, and then it can see that the repeat length checks are redundant and remove them.  And if it fails to, then the length check is cheap (and this is compatibility code anyway).

Reply:
ohh, I missed that. I have now updated it.